OSDOCS-4072: adds ingress firewall node operator (#52114)
openshift/openshift-docs: modules/nw-infw-operator-rules-object.adoc

// Module included in the following assemblies:
//
// * networking/ingress-node-firewall-operator.adoc

:_content-type: CONCEPT
[id="nw-ingress-node-firewall-operator-rules-object_{context}"]
= Ingress node firewall rules object

The fields for the Ingress Node Firewall Operator rules object are described in the following table:

.Ingress node firewall rules object
[cols=".^2,.^2,.^6a",options="header"]
|====
|Field|Type|Description

|`metadata.name`
|`string`

> |The name of the CR object. The name of the config object must be `ingressnodefirewall`.

Reviewer: why we are referring to fw config object while we are talking about fw rules ?
fw rules are cluster scopped resource and you can have many objs

|`interfaces`
|`array`
|The fields for this object specify the interfaces to apply the rules.

|`nodeSelector.matchLabels`
|`array`
|You can use `nodeSelector` `matchLabels` to select the nodes to apply the firewall rules to. Set the value of `do-node-ingress-firewall` set to `true` to apply the rule.
|====

[discrete]
[id="nw-infw-ingress-rules-object_{context}"]
=== `Ingress` object configuration

The values for the `ingress` object are defined in the following table:

> .`ingress` object

Author: Added

[cols=".^3,.^2,.^5a",options="header"]
|====
|Field|Type|Description

|`sourceCIDRs`
|`array`
|You can configure multiple CIDRs from different address families.

[NOTE]
====

> Different CIDRs allow you to use the same order rule. In the case that there are multiple `IngressNodeFirewall` objects for the same nodes and interfaces with overlapping CIDRs, the `order` field will specify which rule is applied first.

Reviewer: that is correct

Different CIDRs allow you to use the same order rule. In the case that there are multiple `IngressNodeFirewall` objects for the same nodes and interfaces with overlapping CIDRs, the `order` field will specify which rule is applied first.
====

|`rules`
|`array`
|Ingress firewall `rules.order` are ordered starting at `1` for each `source.CIDR` with up to 100 rules per CIDR. Lower order rules are executed first.

`protocolConfig.protocol` supports the following protocols: TCP, UDP, SCTP, ICMP and ICMPv6.

[NOTE]
====

> For `protocol.ICMP`, ICMP and ICMPv6 rules can match against ICMP and ICMPv6 type and code. For `TCP`, `UDP`, and `SCTP` rules can match against single DstPort or a range of ports using `<start-end>` format.

Reviewer 2: The actual range would be (start : end-1) so if you define "30000-32767" say, range in effect would be 30000-32766

====

`action` is either allow to apply the rule or deny to disallow the rule. Ingress firewall rules are verified using verfication webhook that fails any invalid configuration. The verification webhook prevents you from blocking any critical cluster ports such as api server or ssh.
|====

[discrete]
[id="nw-ingress-node-firewall-example-cr_{context}"]
== Ingress node firewall rules object example

A complete configuration is specified in the following example:

> .Example Ingress Node Firewall Configuration

Reviewer 3: There is ingress node firewall configuration object vs firewall rules. We should be careful with Configuration word here, make it small case.

> .Example Ingress Node Firewall configuration

Author: synced them up

[source,yaml]
----
apiVersion: ingressnodefirewall.openshift.io/v1alpha1
kind: IngressNodeFirewall
metadata:
  name: ingressnodefirewall
spec:
  interfaces:
  - eth0
  nodeSelector:
    matchLabels:

>       <do-node-ingress-firewall>: 'true'

Reviewer 4: `<do_node_ingress_firewall>`

Underscore for replaceable values in angle brackets per [repo guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output).

kind: IngressNodeFirewall
metadata:
  name: ingressnodefirewall
spec:
  interfaces:
  - eth0
  nodeSelector:
    matchLabels:
      <do_node_ingress_firewall>: 'true'
  ingress:
  - sourceCIDRs:
       - 172.16.0.0/12
    rules:
    - order: 10
      protocolConfig:
        protocol: ICMP
        icmp:
          icmpType: 8 #ICMP Echo request
      action: Deny
    - order: 20
      protocolConfig:
        protocol: TCP
        tcp:
          ports: "8000-9000"
      action: Deny
  - sourceCIDRs:
       - fc00:f853:ccd:e793::0/64
    rules:
    - order: 10
      protocolConfig:
        protocol: ICMPv6
        icmpv6:
          icmpType: 128 #ICMPV6 Echo request
      action: Deny
----

[discrete]
[id="nw-ingress-node-firewall-zero-trust-example-cr_{context}"]
== Zero trust Ingress Node Firewall rules object example

Zero trust Ingress Node Firewall rules are used with multi-interface clusters. If a user wanted to provide additional security, you can use zero trust Ingress Node Firewall rules to drop all traffic on a specific interface except for SSH.

A complete configuration of a zero trust Ingress Node Firewall rules is specified in the following example:

> .Example zero trust Ingress Node Firewall rules

Reviewer 2: * All ports above would be kubelet, etcd, api etc (can be referecned from TE slides)

> .Example zero trust Ingress Node Firewall rules

Author: I added an IMPORTANT callout that tells users to whitelist all ports.

> .Example zero trust Ingress Node Firewall rules

Reviewer 2: Look good, Joe. Thanks.

// Module included in the following assemblies:
//
// * networking/ingress-node-firewall-operator.adoc

:_content-type: CONCEPT
[id="nw-ingress-node-firewall-operator-rules-object_{context}"]
= Ingress Node Firewall rules object

The fields for the Ingress Node Firewall rules object are described in the following table:

.Ingress Node Firewall rules object
[cols=".^2,.^2,.^6a",options="header"]
|====
|Field|Type|Description

|`metadata.name`
|`string`

> |The name of the CR object. The name of the config object must be `ingressnodefirewall`.

Reviewer 3: @JoeAldinger The name of object does not need to be ingressnodefirewall because there can be more than one objects and this CR is for creating multiple firewall rules.

|`interfaces`
|`array`
|The fields for this object specify the interfaces to apply the firewall rules to. For example, `- en0` and
`- en1`.

|`nodeSelector`
|`array`
|You can use `nodeSelector` to select the nodes to apply the firewall rules to. Set the value of your named `nodeselector` labels to `true` to apply the rule.

|`ingress`

> |`object`

Author: I'm going to leave it as is so that it matches some of the other tables in `Networking`.

====
|====

[discrete]
[id="nw-ingress-node-firewall-example-cr_{context}"]
== Ingress Node Firewall rules object example

A complete Ingress Node Firewall configuration is specified in the following example:

.Example Ingress Node Firewall configuration
[source,yaml]
----
apiVersion: ingressnodefirewall.openshift.io/v1alpha1
kind: IngressNodeFirewall
metadata:
  name: ingressnodefirewall
spec:
  interfaces:
  - eth0
  nodeSelector:
    matchLabels:
      <do_node_ingress_firewall>: 'true'
  ingress:
  - sourceCIDRs:
       - 172.16.0.0/12
    rules:
    - order: 10
      protocolConfig:
        protocol: ICMP
        icmp:
          icmpType: 8 #ICMP Echo request
      action: Deny
    - order: 20
      protocolConfig:
        protocol: TCP
        tcp:
          ports: "8000-9000"
      action: Deny
  - sourceCIDRs:
       - fc00:f853:ccd:e793::0/64
    rules:
    - order: 10
      protocolConfig:
        protocol: ICMPv6
        icmpv6:
          icmpType: 128 #ICMPV6 Echo request
      action: Deny
----

[discrete]
[id="nw-ingress-node-firewall-zero-trust-example-cr_{context}"]
== Zero trust Ingress Node Firewall rules object example

> Zero trust Ingress Node Firewall rules are used with multi-interface clusters. If a user wanted to provide additional security, you can use zero trust Ingress Node Firewall rules to drop all traffic on a specific interface except for SSH.

Reviewer 5: ```suggestion
Zero trust Ingress Node Firewall rules can provide additional security to multi-interface clusters. You can use zero trust Ingress Node Firewall rules to drop all traffic on a specific interface except for SSH, for example. 
```
Rewording to avoid passive voice

> A complete configuration of a zero trust Ingress Node Firewall rules is specified in the following example:

Reviewer 5: ```suggestion
A complete configuration of a zero trust Ingress Node Firewall rule set is specified in the following example:
```

[IMPORTANT]
====

> Users need to whitelist all ports their application will use in the following case to ensure proper functionality.

Reviewer 5: ```suggestion
Users need to add all ports their application will use to their allowlist in the following case to ensure proper functionality.
```
